src/protocols/secure_channel/PASESession.cpp

```diff
@@ -578,115 +651,131 @@ CHIP_ERROR PASESession::HandleMsg1_and_SendMsg2(const System::PacketBufferHandle
-    encryptionKeyId = chip::Encoding::LittleEndian::Read16(buf);
-    msg->ConsumeHead(sizeof(encryptionKeyId));
-    buf     = msg->Start();
-    buf_len = msg->DataLength();
+    const uint8_t * Y;
+    size_t Y_len = 0;
 
-    ChipLogDetail(SecureChannel, "Peer assigned session key ID %d", encryptionKeyId);
-    SetPeerKeyId(encryptionKeyId);
+    const uint8_t * peer_verifier;
+    size_t peer_verifier_len = 0;
 
-    err = mSpake2p.ComputeRoundTwo(buf, kMAX_Point_Length, verifier, &verifier_len_raw);
-    SuccessOrExit(err);
-    VerifyOrExit(CanCastTo<uint16_t>(verifier_len_raw), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
-    verifier_len = static_cast<uint16_t>(verifier_len_raw);
+    uint32_t decodeTagIdSeq = 0;
 
-    {
-        const uint8_t * hash = &buf[kMAX_Point_Length];
-        err                  = mSpake2p.KeyConfirm(hash, kMAX_Hash_Length);
-        if (err != CHIP_NO_ERROR)
-        {
-            spake2pErr = Spake2pErrorType::kInvalidKeyConfirmation;
-            SuccessOrExit(err);
-        }
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
-        err = mSpake2p.GetKeys(mKe, &mKeLen);
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    Y_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(Y));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    peer_verifier_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(peer_verifier));
+
+    SuccessOrExit(err = mSpake2p.ComputeRoundTwo(Y, Y_len, verifier, &verifier_len));
+
+    err = mSpake2p.KeyConfirm(peer_verifier, peer_verifier_len);
+    if (err != CHIP_NO_ERROR)
+    {
+        spake2pErr = kProtocolCodeNoSharedRoot;
         SuccessOrExit(err);
     }
 
+    SuccessOrExit(err = mSpake2p.GetKeys(mKe, &mKeLen));
+
     {
-        Encoding::PacketBufferWriter bbuf(System::PacketBufferHandle::New(verifier_len));
-        VerifyOrExit(!bbuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+        size_t data_len = EstimateTLVStructOverhead(verifier_len, 1);
+
+        System::PacketBufferHandle msg3 = System::PacketBufferHandle::New(data_len);
+        VerifyOrExit(!msg3.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+        System::PacketBufferTLVWriter tlvWriter;
+        tlvWriter.Init(std::move(msg3));
 
-        bbuf.Put(verifier, verifier_len);
-        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+        SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+        SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), ByteSpan(verifier, verifier_len)));
```

Comment:
Same comment as for Pake2: suggest replacing `(1)` in `TLV::ContextTag(1)` with something that expresses which field name is being set.

Overall, I think it's OK to use 1,2,3 directly in ContextTag (I find it makes it much easier to follow), but that's only OK if the field being written has the right name. The other approach is to rename `verifier` here to `cA`...

Reply:
Added a constant for the ContextTag value.